oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
 	 *
 	 * @param allowUriQueryParameter if the URI query parameter is supported
 	 */
-	public void setAllowUriQueryParameter(boolean allowUriQueryParameter) {
+	public void setAllowUriQueryParameter(final boolean allowUriQueryParameter) {
 		this.allowUriQueryParameter = allowUriQueryParameter;
 	}
 
-	private static String resolveFromAuthorizationHeader(HttpServletRequest request) {
-		String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
-		if (StringUtils.hasText(authorization) && authorization.startsWith("Bearer")) {
-			Matcher matcher = authorizationPattern.matcher(authorization);
+	private void validate(final HttpServletRequest request, final String authorizationHeaderToken) {
+		if (hasMultipleAccessTokens(request, authorizationHeaderToken)) {
+			throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MULTIPLE);
+		}
+		if (this.allowFormEncodedBodyParameter == true) {
+			if (this.allowUriQueryParameter == false && isTokenInUriQueryParameter(request)) {
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
+			}
+		}
+	}
 
+	private String resolveFromAuthorizationHeader(final HttpServletRequest request) {
+		final String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
+		if (authorization != null && authorization.startsWith("Bearer")) {
+			final Matcher matcher = AUTHORIZATION_PATTERN.matcher(authorization);
 			if (!matcher.matches()) {
-				BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_TOKEN,
-						HttpStatus.UNAUTHORIZED,
-						"Bearer token is malformed",
-						"https://tools.ietf.org/html/rfc6750#section-3.1");
-				throw new OAuth2AuthenticationException(error);
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
 			}
-
 			return matcher.group("token");
 		}
 		return null;
 	}
 
-	private static String resolveFromRequestParameters(HttpServletRequest request) {
-		String[] values = request.getParameterValues("access_token");
-		if (values == null || values.length == 0)  {
+	private String resolveFromRequestParameters(final HttpServletRequest request) {
+		// By this point there is either zero or one value
+		final String value = request.getParameter(ACCESS_TOKEN_PARAM_NAME);
+		if (value != null && isParameterTokenSupportedForRequest(request)) {
+			final Matcher matcher = JWT_PATTERN.matcher(value);
+			if (!matcher.matches()) {
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
+			}
+			return value;
+		} else {
 			return null;
 		}
+	}
 
-		if (values.length == 1) {
-			return values[0];
-		}
+	private boolean isParameterTokenSupportedForRequest(final HttpServletRequest request) {
+		return this.allowUriQueryParameter || (this.allowFormEncodedBodyParameter
```

Comment:
If I'm not missing something, this now allows the use of `allowUriQueryParameter` for all request methods - I don't like that and believe the use of `allowUriQueryParameter` should be limited to `GET` as it originally was. The point is if one can do `POST` then they can provide the parameter as `application/x-www-form-urlencoded` - query parameter support should be only supported as fallback for challenged clients that cannot do anything other than `GET`.

Reply:
Specification does not restrict us to from sending access token in query parameter for POST request

Reply:
The use of query parameter is clearly discouraged by the spec, and it's there just in cases where _it is impossible to transport the access token in the "Authorization" request header field or the HTTP request entity-body_. Having in mind the HTTP semantics, while not explicitly stated, this is clearly targeted at client incapable of doing anything other than a `GET` request. There's no valid reason not to use form-encoded body parameter if client is capable of doing a `POST` request.

Reply:
It is better to stick to the specifications or we will build solution that is incompatible for some scenarios. For example if we limit only to `GET` method then we are also prohibiting `HEAD` request which is not correct. Let's not add extra restrictions that are not mentioned in specs unless it is required.

I understand "use of query parameter is clearly discouraged by the spec" and that is the reason why we have `allowUriQueryParameter = false` by default.

Reply:
Agreed that if the client can send a POST, then they can send a POST body. This means that we are within the spec to disallow a query parameter in that case (since it isn't impossible, as @vpavic explained). @gburboz do you have an example where this would not be the case?

This also has the happy consequence that we are not enabling bad behavior, as @rwinch noted.

Also, we can typically relax restrictions iteratively (e.g. in future PRs). And folks can very easily supply their own implementation for their corner case. If it becomes clear that this flexibility is necessary, we can add it with relative ease.

With these items in mind, let's only check for the query parameter when it is a GET request. I like this, too, because it narrows the focus of this PR closer to [its original goal](https://github.com/spring-projects/spring-security/issues/5668).

Reply:
@jzheaux , below are few examples of query param for method other than `GET` while client is unable to create form body
* `HEAD` method which could be used when `GET` request is processing heavy or results in too much data
* `TRACE` method which can be used for debugging issues with `GET` method
* `POST`/`PUT`/`PATCH`/`DELETE` method but data is sent as query params (non `GET` method used here is to support REST paradigm)

@vpavic since everyone in this PR are leaning towards not supporting scenarios mentioned above, I have reverted the change and limited query param only to `GET` method. In light of above example if we feel change is required, we can raise new issue and raise different PR against it.